### ISSUE TITLE
NameError: uninitialized constant TTY::Prompt::Utils

### ISSUE THREAD

**Issue by ioquatix** [piotrmurach/tty-prompt#37]
Gemfile

```
gem 'tty-prompt'
```

Code

```
require 'tty/prompt'
prompt = TTY::Prompt.new
```

Error

```
NameError: uninitialized constant TTY::Prompt::Utils
/home/samuel/.rvm/gems/ruby-2.3.2/gems/tty-prompt-0.10.0/lib/tty/prompt.rb:69:in `initialize'
```

**Comment by ioquatix** [comment]
I took a look at the code and it was because I wrote `require 'tty/prompt` rather than `require 'tty-prompt'`. I suggest that you fix this.

It should be possible to include any individual file and have it require_relative all it's dependencies. This is good Ruby.

**Comment by piotrmurach** [comment]
Thanks for using the library. 

I cannot make judgment on what is or is not a good Ruby in this case as even Ruby core contributors seem to have different opinions on the issue - [see](https://bugs.ruby-lang.org/issues/8781). Rubyists generally have different approaches to loading files - [see](https://www.reddit.com/r/ruby/comments/3341zm/require_best_practices/).

I'm relying on the files being added to load path when rubygems requires them. The entry point for this library is `tty-prompt` as the usual convention is to require gems by their names. However, I'm looking forward to PR that allows to load this library using the `tty/prompt` loading path.

**Comment by ioquatix** [comment]
If you don't agree that's fine.

**Comment by piotrmurach** [comment]
I cannot see anywhere in what I wrote that I don't agree with you. I simply pointed out the reason behind my choice of structuring the codebase, there are good and bad sides to each solution. I'm happy to make it work for `tty/prompt` and provided you with an opportunity to contribute to this project as I do welcome new contributors with open arms. I do not have access to the internet at home and it would be hard to get this done for me at the moment.

**Comment by ioquatix** [comment]
`require_relative` is faster and predictable. It's a nice way to structure code. There is no reason to use anything else within the boundaries of a single library. Between libraries, you should of course use `require`. `require_relative` is ridiculously slow compared to `require` in larger projects.

In Ruby, I'm of the option that `-` should not appear in file names except in very specific cases.

In general, a class named `MyLibrary::MyClass` would belong in a gem called `my-library` and in a path `lib/my_library/my_class.rb`. So, humbly, I believe you've done a good job of naming the internal code and gems. Some people would have called it `Tty::Prompt` but I think that's fugly.

Honestly, I think this [ruby naming conventions] is actually a bit unfortunate. When I write C++, I make the path name identical to the namespace/class hierarchy. So, a class called `MyLibrary::MyClass` in C++ would be in the file `source/MyLibrary/MyClass.hpp`. Honestly, I think this is simpler and better overall since there can be no confusion.

Anyway, I'd love to submit a PR. But your `tty` project is pretty huge and I don't have time to make all the changes. Even if I did, I wouldn't do it unless you were 100% onboard with the idea, otherwise it's just a waste of my time. I'm not trying to be offensive here, just matter of fact.

I bet that you'd be best to do it yourself and see for yourself that it makes the code better.

**Comment by piotrmurach** [comment]
Thank you for your comment. I agree with your reasoning, it makes perfect sense. The one thing I always liked about `require` over `require_relative` was the flexibility of placing the files whenever. I have just recently stopped creating a single file that requires all the dependencies in favour of each single file requiring only the dependencies and files it needs. It really does help with maintenance and structure of files even though things may be required in more than one place. Regardless of what technique is used I will see if I can make this library work for both `tty-prompt` and `tty/prompt` requires.

**Comment by ioquatix** [comment]
> The one thing I always liked about require over require_relative was the flexibility of placing the files whenever.

That's a surprising, IMHO. For me, it's the exact opposite, `require_relative` forces an implicit contract on the location of the file, and it makes it easier to know where to find something.

> I have just recently stopped creating a single file that requires all the dependencies in favour of each single file requiring only the dependencies and files it needs. It really does help with maintenance and structure of files even though things may be required in more than one place.

For me, it's also about the performance of running individual tests in large projects. It requires the minimum set of files required for the test to work.

**Comment by piotrmurach** [comment]
Thanks for this discussion I went ahead with `require_relative` change, after reading around the subject it makes perfect sense to use it. It will take some time to update all other libraries, but slowly I should get there.

**Comment by ioquatix** [comment]
Yay, I look forward to trying it out :)

**Comment by piotrmurach** [comment]
Cheers, this is gonna be important release that hopefully makes this library work smoothly like silk on Windows :shipit:

**Comment by piotrmurach** [comment]
Released `v0.11.0`. Enjoy!
